### PR TITLE
feat(cache): add entity caching

### PR DESCRIPTION
**Pull request by xbaun**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/datorama/akita/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the new behavior?

Entities of an entity store are individually cached.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by xbaun** [comment]
@NetanelBasal A first proposal to implement entity caching independently of the store. I am not yet sure which query selectors are suitable. The tests also need to be expanded. The cache itself is implemented as a priority scheduler and should therefore be resource efficient, even with a large number of entities.

File: libs/akita/src/__tests__/entityStoreCache.spec.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    );
+
+    store.set([
+      { id: 1, token: undefined },
+      { id: 2, token: 'token 2' },
+    ]);
+
+    expect(store._value().ids.length).toBe(2);
+
+    {
```
Comment (NetanelBasal) [inline libs/akita/src/__tests__/entityStoreCache.spec.ts:144]: Where do you test that it doesn't invoke the source when it's cached?

File: libs/akita/src/__tests__/queryEntity.spec.ts [hunk 2]
```diff
@@ -383,48 +383,48 @@ describe('Entities Query', () => {
...
+          expect(spy).toHaveBeenCalledTimes(3);
           expect(spy).toHaveBeenCalledWith(false);
           store.setHasCache(true);
-          expect(spy).toHaveBeenCalledTimes(8);
-          expect(spy).toHaveBeenCalledWith(true);
-          store.setHasCache(true);
-          expect(spy).toHaveBeenCalledTimes(8);
+          expect(spy).toHaveBeenCalledTimes(4);
         });
       });
```
Comment (NetanelBasal) [inline libs/akita/src/__tests__/queryEntity.spec.ts:411]: What we're going to do with the commented specs?

File: libs/akita/src/__tests__/setLoading.spec.ts [hunk 3]
```diff
@@ -28,13 +28,13 @@ describe('setLoading', () => {
     expect(query.getValue().loading).toBe(false);
     cacheable(store, request).subscribe(spy);
     // from inside withLoading
-    expect(store.setLoading).toBeCalledTimes(2);
```
Comment (NetanelBasal) [inline libs/akita/src/__tests__/setLoading.spec.ts:31]: Why the change?

File: libs/akita/src/lib/entityCacheOperators.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  TEntity = getEntityType<TState>,
+  TIdType = getIDType<TState>
+>(store: TStore, id: TIdType): MonoTypeOperatorFunction<TEntity> {
+  const config = store.cacheConfig;
+
+  if (config === undefined || config.ttl === undefined) {
+    throw Error(`No cache ttl in @StoreConfig set of '${store.storeName}' to compute entity expiration: ` + `@StoreConfig({ cache: { ttl: 1000 }})`);
+  }
+
+  const result = store
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityCacheOperators.ts:111]: Can we make it DRY, please? I mean extract to a function and use it in both operators, please.

File: libs/akita/src/lib/entityStore.ts [hunk 5]
```diff
@@ -58,6 +64,19 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
 
   constructor(initialState: Partial<S> = {}, protected options: Partial<StoreConfigOptions> = {}) {
     super({ ...getInitialEntitiesState(), ...initialState }, options);
+
+    this.ttlQueue.expired$.pipe(filter((ttl) => ttl.type === TTLType.Entity)).subscribe((ttl) => {
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:68]: Extract to a function and call it only if users enable the caching feature.

File: libs/akita/src/lib/entityStore.ts [hunk 6]
```diff
@@ -58,6 +64,19 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
...
+      if (ttl.id && this.ids.includes(ttl.id)) {
+        this._setState((state) => {
+          const idsExpired = state.idsExpired || [];
+
+          return {
+            ...state,
+            idsExpired: idsExpired.includes(ttl.id) ? idsExpired : [...idsExpired, ttl.id],
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:75]: Do you think we need to add the property to the state? Another option is to add it to the instance.

File: libs/akita/src/lib/entityStore.ts [hunk 7]
```diff
@@ -230,6 +261,12 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
     }
 
     this.entityActions.next({ type: EntityActions.Update, ids });
+
+    const ttlConfig = this.getCacheTTL();
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:265]: Make all of the same usages DRY, please.

File: libs/akita/src/lib/store.ts [hunk 8]
```diff
@@ -299,6 +303,9 @@ export class Store<S = any> {
       this._initialState = initialState;
     }
     isDev() && assertStoreHasName(this.storeName, this.constructor.name);
+    this.ttlQueue.expired$.pipe(filter((ttl) => ttl.type === TTLType.Store)).subscribe(() => {
```
Comment (NetanelBasal) [inline libs/akita/src/lib/store.ts:306]: Extract to a method and call if only if users enabled the cache feature, please.

File: libs/akita/src/lib/entityStore.ts (hunk 6 above)
Comment (xbaun) [reply libs/akita/src/lib/entityStore.ts:75]: If it's part of the state it can be subscribed through `select` without introducing a new observable. In addition, expirations can be debugged and tracked through devtools. So I thought it would be fine to put it in the state itself.

File: libs/akita/src/__tests__/setLoading.spec.ts (hunk 3 above)
Comment (xbaun) [reply libs/akita/src/__tests__/setLoading.spec.ts:31]: The current behaviour is, that `EntityStore.set()` set `hasCache` of the store to true. I remove it, because entities are treated separately from the store: https://github.com/datorama/akita/pull/471/files#diff-be97ad95765300e4c32ef78919456d39R141

The `setLoading` call count changed as a result of `cacheable` is triggered differently because `hasCache` isn't set any more.

File: libs/akita/src/__tests__/queryEntity.spec.ts (hunk 2 above)
Comment (xbaun) [reply libs/akita/src/__tests__/queryEntity.spec.ts:411]: Some of them are obsolete and should be removed. Others must be adapted to cover the new behaviour. I have commented them out to review them again when all changes have been made.

File: libs/akita/src/__tests__/entityStoreCache.spec.ts (hunk 1 above)
Comment (xbaun) [reply libs/akita/src/__tests__/entityStoreCache.spec.ts:144]: The code block below: https://github.com/datorama/akita/pull/471/files/8f24df0e47610bafbe4735e095d7d4c3dd074ba1#diff-de1efd0295ddb2cc00b19d5c3f6ec4f4R155

File: libs/akita/src/lib/entityStore.ts [hunk 9]
```diff
@@ -660,6 +692,37 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
...
+
+  protected watchCacheTTL() {
+    super.watchCacheTTL();
+    this.ttlQueue.expired$.pipe(filter((ttl) => ttl.type === TTLType.Entity)).subscribe((ttl) => {
+      if (ttl.id && this.ids.includes(ttl.id)) {
+        this._setState((state) => {
+          const idsExpired = state.idsExpired || [];
+          return {
+            ...state,
+            idsExpired: idsExpired.includes(ttl.id) ? idsExpired : [...idsExpired, ttl.id],
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:720]: This will trigger a redundant dispatch in case of `idsExpired.includes(ttl.id)`.

File: libs/akita/src/lib/types.ts [hunk 10]
```diff
@@ -43,10 +45,10 @@ export type IDS = ID | ID[];
 export type PreAddEntity<Entity> = (entity: Entity) => Entity;
 export type PreUpdateEntity<Entity> = (prevEntity: Entity, nextEntity: Entity) => Entity;
 
-export type StoreCache = {
-  active: BehaviorSubject<boolean>;
-  ttl: number;
-};
+// export type StoreCache = {
```
Comment (NetanelBasal) [inline libs/akita/src/lib/types.ts:48]: Don't forget about this one.

File: libs/akita/src/lib/updateEntities.ts [hunk 11]
```diff
@@ -90,5 +90,6 @@ export function updateEntities<S extends EntityState<E>, E>({ state, ids, idKey,
       ...updatedEntities,
     },
     ids: updatedIds,
+    idsExpired: coerceArray(state.idsExpired).filter((id) => !updatedIds.includes(id)),
```
Comment (NetanelBasal) [inline libs/akita/src/lib/updateEntities.ts:93]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/addEntities.ts [hunk 12]
```diff
@@ -36,11 +37,12 @@ export function addEntities<S extends EntityState<E>, E>({ state, entities, idKe
           ...state,
           entities: {
             ...state.entities,
-            ...newEntities
+            ...newEntities,
           },
-          ids: options.prepend ? [...newIds, ...state.ids] : [...state.ids, ...newIds]
+          ids: options.prepend ? [...newIds, ...state.ids] : [...state.ids, ...newIds],
+          idsExpired: coerceArray(state.idsExpired),
```
Comment (NetanelBasal) [inline libs/akita/src/lib/addEntities.ts:43]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/entityStore.ts [hunk 13]
```diff
@@ -580,6 +601,17 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
     return this.ui;
   }
 
+  /**
+   * Reset entity cache: Cancel all TTLs and clear expired entity ids.
+   */
+  resetTTLEntityCache() {
+    this.ttlQueue.reset();
+    this.update((state) => ({
+      ...state,
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:610]: You can omit the `...state`

File: libs/akita/src/lib/entityStore.ts [hunk 14]
```diff
@@ -335,15 +344,21 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
     this._setState((state) => ({
       ...state,
       ids: addedIds.length ? [...state.ids, ...addedIds] : state.ids,
+      idsExpired: coerceArray(state.idsExpired).filter((id) => !updatedIds.includes(id)),
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:347]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/entityStore.ts [hunk 15]
```diff
@@ -151,6 +158,7 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
         this.handleUICreation(true);
       }
 
+      this.updateEntityTTL(data.newIds);
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:161]: if (this.hasCacheTTLEnabled()) {

File: libs/akita/src/lib/entityStore.ts [hunk 16]
```diff
@@ -114,6 +120,7 @@ export class EntityStore<S extends EntityState = any, EntityType = getEntityType
       this.handleUICreation();
     }
 
+    this.updateEntityTTL(this.ids);
```
Comment (NetanelBasal) [inline libs/akita/src/lib/entityStore.ts:123]: if (this.hasCacheTTLEnabled()) {

File: libs/akita/src/lib/removeEntities.ts [hunk 17]
```diff
@@ -22,7 +23,8 @@ export function removeEntities<S extends EntityState<E>, E>({ state, ids }: Remo
   const newState = {
     ...state,
     entities: newEntities,
-    ids: state.ids.filter(current => ids.includes(current) === false)
+    ids: state.ids.filter((current) => !ids.includes(current)),
+    idsExpired: coerceArray(state.idsExpired).filter((id) => !ids.includes(id)),
```
Comment (NetanelBasal) [inline libs/akita/src/lib/removeEntities.ts:27]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/removeEntities.ts [hunk 18]
```diff
@@ -38,6 +40,7 @@ export function removeAllEntities<S>(state: StateWithActive<S>): S {
     ...state,
     entities: {},
     ids: [],
-    active: isMultiActiveState(state.active) ? [] : null
+    idsExpired: [],
```
Comment (NetanelBasal) [inline libs/akita/src/lib/removeEntities.ts:43]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/setEntities.ts [hunk 19]
```diff
@@ -43,14 +43,15 @@ export function setEntities<S extends EntityState<E>, E>({ state, entities, idKe
...
     // it's an object
     newEntities = isNativePreAdd ? entities : applyMiddleware(entities, preAddEntity);
-    newIds = Object.keys(newEntities).map(id => (isNaN(id as any) ? id : Number(id)));
+    newIds = Object.keys(newEntities).map((id) => (isNaN(id as any) ? id : Number(id)));
   }
 
   const newState = {
     ...state,
     entities: newEntities,
+    idsExpired: [],
```
Comment (NetanelBasal) [inline libs/akita/src/lib/setEntities.ts:52]: We shouldn't add this if the user doesn't enable this feature.

File: libs/akita/src/lib/entityStore.ts (hunk 16 above)
Comment (xbaun) [reply libs/akita/src/lib/entityStore.ts:123]: It's checked within `updateEntityTTL()`
